crates/sui-core/src/consensus_handler.rs

```diff
@@ -96,29 +100,21 @@ impl<T, C> ConsensusHandler<T, C> {
     /// Updates the execution indexes based on the provided input. Some is returned when the indexes
     /// are updated which means that the transaction has been seen for first time. None is returned
     /// otherwise.
-    fn update_hash(
-        &mut self,
-        index: ExecutionIndices,
-        v: &[u8],
-    ) -> Option<ExecutionIndicesWithHash> {
-        if let Some(execution_indexes) = update_hash(&self.last_seen, index, v) {
-            self.last_seen = execution_indexes.clone();
-            return Some(execution_indexes);
-        }
-        None
+    fn update_index_and_hash(&mut self, index: ExecutionIndices, v: &[u8]) -> bool {
+        update_index_and_hash(&mut self.last_consensus_stats, index, v)
     }
 }
 
-fn update_hash(
-    last_seen: &ExecutionIndicesWithHash,
+fn update_index_and_hash(
+    last_consensus_stats: &mut ExecutionIndicesWithStats,
     index: ExecutionIndices,
     v: &[u8],
-) -> Option<ExecutionIndicesWithHash> {
-    if last_seen.index >= index {
-        return None;
+) -> bool {
+    if last_consensus_stats.index >= index {
+        return false;
```

Comment:
Is this even necessary, now that the consensus handler is atomic?

Reply:
Missed the comment before merging. This if block could be unnecessary, arguably along with some of the consensus_message_processed, processed_set & processed_cache logic. This can be done in a separately cleanup, because it seems riskier.